genetlink/src/resolver.rs

```diff
@@ -0,0 +1,160 @@
+                    match rx_packet.payload {
+                        NetlinkPayload::InnerMessage(genlmsg) => {
+                            let family_id = genlmsg
+                                .payload
+                                .nlas
+                                .iter()
+                                .find_map(|nla| {
+                                    if let GenlCtrlAttrs::FamilyId(id) = nla {
+                                        Some(*id)
+                                    } else {
+                                        None
+                                    }
+                                })
+                                .ok_or_else(|| {
+                                    GenetlinkError::AttributeNotFound(
+                                        "CTRL_ATTR_FAMILY_ID".to_owned(),
+                                    )
+                                })?;
+
+                            self.cache.insert(family_name, family_id);
+                            return Ok(family_id);
+                        }
+                        NetlinkPayload::Error(e) => return Err(e.into()),
+                        _ => (),
+                    }
+                }
+
+                Err(GenetlinkError::NoMessageReceived)
+            })
+        }
+    }
+
+    pub fn clear_cache(&mut self) {
+        self.cache.clear();
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::new_connection;
+    use std::io::ErrorKind;
+
+    #[tokio::test]
+    async fn test_resolver_nlctrl() {
+        let (conn, handle, _) = new_connection().unwrap();
+        tokio::spawn(conn);
+
+        let mut resolver = Resolver::new();
+        // nlctrl should always be 0x10
+        let nlctrl_fid = resolver.query_family_id(&handle, "nlctrl").await.unwrap();
+        assert_eq!(nlctrl_fid, 0x10);
+    }
+
+    const TEST_FAMILIES: &[&str] = &[
+        "devlink",
+        "ethtool",
+        "acpi_event",
+        "tcp_metrics",
+        "TASKSTATS",
+        "nl80211",
+    ];
+
+    #[tokio::test]
+    async fn test_resolver_cache() {
+        let (conn, handle, _) = new_connection().unwrap();
+        tokio::spawn(conn);
+
+        let mut resolver = Resolver::new();
+
+        // Test if family id cached
+        for name in TEST_FAMILIES.iter().copied() {
+            let id = resolver
+                .query_family_id(&handle, name)
+                .await
+                .or_else(|e| {
+                    if let GenetlinkError::NetlinkError(io_err) = &e {
+                        if io_err.kind() == ErrorKind::NotFound {
+                            // Ignore non exist entries
+                            Ok(0)
+                        } else {
+                            Err(e)
+                        }
+                    } else {
+                        Err(e)
+                    }
+                })
+                .unwrap();
+            if id == 0 {
+                eprintln!(
```

Comment:
Please use `warn!()` instead.

Reply:
It is inside the unit test, I don't think that it should use `warn!()` instead.
Unless you mean using the following work around:
```rust
#[cfg(test)]
use std::{println as info, println as warn};
```

Reply:
Oh. I missed the unit test part. Yes, you can use print in unit test.